lib/grpnames.gi

```diff
@@ -20,46 +20,575 @@
+        UniteSet(Df, H);
+      else
+        DfNs := Filtered(Ns, N ->IsSubset(H, N));
+        gs := Set(DfMinNs, N -> GeneratorsOfGroup(N)[1]);
+        # normal subgroup containing all minimal subgroups cannot have complement in H
+        DfNs := Filtered(DfNs, N -> not IsSubset(N, gs));
+        UniteSet(Df, DirectFactorsOfGroupFromList(H, DfNs, DfMinNs));
+      fi;
+    od;
+    return Df;
+  end);
 
-  function ( G )
+InstallMethod(DirectFactorsOfGroup, "for direct products", true,
+                      [ IsGroup and HasDirectProductInfo ], 10,
+
+  function(G)
+    local i, info, Ns;
+
+    Ns := [];
+    info := DirectProductInfo(G).groups;
+    for i in [1..Length(info)] do
+      UniteSet(Ns, DirectFactorsOfGroup(Image(Embedding(G,i),info[i])));
+    od;
+    return Ns;
+  end);
 
-    local  N, facts, sizes, i, j, s1, s2;
+InstallMethod(DirectFactorsOfGroup, "if normal subgroups are computed", true,
+                      [ IsGroup and HasNormalSubgroups ], 10,
+
+  function(G)
+    local Ns, MinNs, GGd, gs;
+
+    Ns := NormalSubgroups(G);
+    MinNs := MinimalNormalSubgroups(G);
+
+    if Length(MinNs)= 1 then
+      # size of MinNs is an upper bound to the number of components
+      return [ G ];
+    fi;
+
+    if IsSolvableGroup(G) then
+      GGd := CommutatorFactorGroup(G);
+      if IsCyclic(GGd) and IsPrimePowerInt(Size(GGd)) then
+        # G is direct indecomposable, because has a unique maximal subgroup
+        return [ G ];
+      fi;
+    elif Length(MaximalNormalSubgroups(G))= 1 then
+      # size of MaximalNormalSubgroups is an upper bound to the number of
+      # components
+      return [ G ];
+    fi;
+
+    gs := Set(MinNs, N -> GeneratorsOfGroup(N)[1]);
+    # normal subgroup containing all minimal subgroups cannot have complement
+    Ns := Filtered(Ns, N -> not IsSubset(N, gs));
+
+    return DirectFactorsOfGroupFromList(G, Ns, MinNs);
+  end);
+
+InstallMethod(DirectFactorsOfGroup, "generic method", true,
+                        [ IsGroup ], 0,
+  function(G)
+
+    local p,        # prime
+          Gd,       # G'
+          GGd,      # G/G'
+          C,        # Center(G)
+          GC,       # G/C
+          avoid,    # set of elements or indices to avoid
+          Ns,       # list of normal subgroups
+          MinNs,    # list of minimal normal subgroups
+          gs,       # list containing one generator for each MinNs
+          g,        # group element
+          N,        # (possible) component of G, containing g
+          B;        # (possible) complement of G in N
+
+    # for Abelian groups return the canonical decomposition
+    if IsTrivial(G) then
+      return [G];
+    elif IsAbelian(G) then
+      Ns := List(IndependentGeneratorsOfAbelianGroup(G),
+                                                  g -> SubgroupNC(G, [g]));
+      # for infinite groups Set may be very inefficient
+      # might cause problems because for finite groups we return a set
```

Comment:
There is a new function called UnionIfCanEasilySortElements, which returns either the union or conatenation of its arguments (which are supposed to be lists), depending on whether for all of their all elements CanEasilySortElements holds. 

I am not entirely sure, though, that this is the proper library file such a function should be put into.